### ISSUE TITLE
Fails to invoke `after_remove` callback when deleting associated record

### ISSUE THREAD

**Issue by ekampp** [rails/rails#14365]
I have a class with a `has_many` association. I have a `after_add` and an `after_remove` callback on the association, like this:

``` rails
# rails 4.2.0.alpha
has_many :character_levels,
  after_add: :log_adding_event,
  after_remove: :log_removing_event

def log_adding_event
  Log.create!({ event: "adding" })
end

def log_removing_event
  Log.create!({ event: "removing" })
end
```

When adding a character level, I can see that a Log instance is created. I can even replicate in tests. But when deleting an associated record, the callback isn't getting executed correctly:

``` ruby
Character.find(1).character_levels << CharacterLevel.find(1) 
# => Results in a Log record being created

Character.find(1).character_levels.delete(CharacterLevel.find(1))
# => Should result in a Log being created, but doesn't
```

**Comment by jetheredge** [comment]
This is probably not surprising, but this also affects 'has_many through'. I have a 'has_many through' which I was calling 'clear' on to remove the relationships. In the past this would call before_remove on the 'has_many through' relationship but this is no longer working.

**Comment by DmitryBochkarev** [comment]
Any progress? I can confirm this bug on ruby-2.1.2 and rails 4.2.1.

**Comment by ekampp** [comment]
I have moved to Rails 5.x, and found another way of solving my problem, but from a quick and dirty test, it doesn't seem to fixed there either.

**Comment by trayo** [comment]
I've been digging around in ActiveRecord and it doesn't seem to be a bug, it seems to be more of an non-included feature.

Using [zenati's test](https://github.com/rails/rails/issues/14365#issuecomment-54271422) it seems that the Line model doesn't get the same destroy callbacks when you use the `after_remove` option.

Using @sebasoga's solution works because the `destroy` method is being called on the `subject` model that inherits from the `collection_association`. Calling `destroy` on `ticket` goes through a different association. 

That's my two cents of what I found. I'm experimenting with some code to trigger the callbacks, but haven't made much progress.

**Comment by ekampp** [comment]
:+1:

**Comment by kelso** [comment]
This issue is still present in rails 4.2.3

    class Post < ActiveRecord::Base
      has_and_belongs_to_many :car_brands, after_remove: :calculate_posts_count

      def calculate_posts_count(car_brand)
        car_brand.update posts_count: car_brand.posts.count
      end

      before_destroy { |post| post.car_brands.clear }
    end

If I destroy post, where car_brands.clear is called, after_remove on association is not triggered.

**Comment by CHENGP618** [comment]
It's probably not a bug. try to reload the instance

**Comment by dsounded** [comment]
Still present in Rails 5, I can workaround this by calling `after_destroy` on dependent model, but this is not OK, since I need to move my callback method to public zone

**Comment by JerryGreen** [comment]
Guys! Have the same shit. Found a simple and good solution!

My old code (with this bug):
```
  has_many :user_organizations, dependent: :destroy
  has_many :organizations, through: :user_organizations
```

On destroying user_organization its `after_destroy` callback not calling. Sadly. But by adding `dependent: :destroy` to `has_many` with `through` like this:

```
  has_many :user_organizations, dependent: :destroy
  has_many :organizations, through: :user_organizations, dependent: :destroy
```

It solves the problem. `dependent: :destroy` for `has_many :organizations` with `through` parameter works not really as expected. It changes the way of removing associated objects. It forces to use `destroy` instead of `delete` so callbacks are called.

**Comment by chris-roerig** [comment]
I'm on Rails 4.2.8. Noticed the lambda does not fire but referencing a method does work. 

Assuming the removal is performed as such:

```
user.delete(foo)
```

do_something is not called

```
# (not working)
has_many :foos, through: :foos_bars, after_remove: -> { do_something }
```

However, this works

```
# (working)
has_many :foos, through: :foos_bars, after_remove: :after_remove_do_something

def after_remove_do_something(assoc)
  do_something
end
```

**Comment by SampsonCrowley** [comment]
3 years later, on rails 5 this is still an issue

```ruby
class User
  has_many :payments, inverse_of: :user, after_add: :create_traveler_or_update_balance, after_remove: :set_balance
end

class Payment < ApplicationRecord
  after_commit :send_receipt, on: :create
  belongs_to :user, inverse_of: :payments
  attribute :amount, :money_column

  def send_receipt
    PaymentMailer.receipt(self.id).deliver_later
  end
end

# destroying a payment never calls the `set_balance` method:
user.payments.last.destroy

# yet, creating a payment always runs `create_traveler_or_update_balance`:
user.payments.create(payment_attributes)
```

**Comment by ekampp** [comment]
First off, woah(!), I can't believe I raised this three years ago...

@SampsonCrowley according to @chris-roerig referencing the after_remove method by name, like you do works, but then he's on Rails 4.2.8, and you seem to be on Rails 5.x. So perhaps there's a regression there somewhere. Just mentioning this in case someone felt like looking into this.

**Comment by SampsonCrowley** [comment]
@ekampp i can say with a surety that referencing by name does not work in rails `5.0.6`... i rarely use procs/lambdas for callback functions so if named functions work in `4` it is definitely regressed back to not working (though it seems more likely that something else was calling that named function, as i would assume the way that callbacks are actually executed is the same for all callbacks, and only the triggers change, if it was actually being triggered at all, it seems unlikely the lambda vs named function would be relevant but i could be wrong)

**Comment by JanStevens** [comment]
I can also confirm that this is a regression in rails 5. Updating our app from rails 4 to 5 this is a roadblock as we heavily depend on the functionality.

We have a `has_many through` association with the `after_add` and `after_remove` callback. I tried adding the `after_remove` to the main association and has through association nothing seems to work.

**Comment by rubendinho** [comment]
it looks like to trigger the `after_remove` callback you have to explicitly do `object.collection.delete(associated_object)`. Deleting the parent object or deleting the associated object directly do not trigger the `after_remove` callback. 

Is this by design?

**Comment by bnwpro** [comment]
@JerryGreen's fix worked for me.  In my case, I have an invite/uninvite Mailer set to send when a **Membership** association is created/destroyed through a polymorphic Group model:

```
  Module Invitable
  after_save -> { invite_to_group }     
  after_commit -> { uninvite_from_group }, on: :destroy
  ...
```

The association model:

```
  Class Membership
  has_many :memberships, as: :memberable, dependent: :destroy     
  has_many :members, through: :memberships, source: :user     
  has_many :users, -> { distinct }, through: :memberships
  ...
```

When the Group record is created/destroyed, the **Membership** association callbacks work as expected.
When the Parent model is updated, the association callback for `after_save -> { invite_to_group }` works, however, `{ uninvite_from_group }` doesn't get called.

Looking at my logs, and only during the removal of associated records during the **Update** action, the destroy method is being called via mass assignment directly on the DB and bypassing the Membership model thus never triggering the `after_commit -> { uninvite_from_group }` callback.

So, applying @JerryGreen's workaround I added the additional `dependent: :destroy` to **Membership**:

 ```
  Class Membership
  has_many :memberships, as: :memberable, dependent: :destroy     
  has_many :members, through: :memberships, source: :user     
  has_many :users, -> { distinct }, through: :memberships, dependent: :destroy <--
  ...
```

..and all is well.

Hope it helps someone!

**Comment by MatayoshiMariano** [comment]
The same thing happens in rails 5 with mongoid

**Comment by nickborromeo** [comment]
👋 everyone, I dug into this a little bit today and I think I am leaning towards this not being a but and maybe can benefit from a little more clarification in the docs. 

TLDR; I think we can go ahead and close this issue due this not being a bug but more of an incorrect expectation when removing associations. I can try to follow up on updating the docs based on the details in this comment. 

cc @ekampp @zenati @rafaelfranca 

First thing we have to clarify that this is an [`Association Callback`](https://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html) not an [`ActiveRecord Callback`](https://api.rubyonrails.org/classes/ActiveRecord/Callbacks.html) they are a little different and could explain the behavior that is being shown here. 

I found a test in rails that we can use as an example
https://github.com/rails/rails/blob/b368511db24fb126545327984511281153730ec4/activerecord/test/cases/associations/callbacks_test.rb#L105-L111

### Why this test works?
The association is defined with `dependent: :destroy`, this will result in calling `handle_dependency` in the `HasManyAssociation` class. `:destroy` will then call `destroy_all`

https://github.com/rails/rails/blob/3e970d09802600e68bdcc9e385f99d173e549c66/activerecord/lib/active_record/associations/has_many_association.rb#L25-L28

https://github.com/rails/rails/blob/3dff30c14fa25f211088d62da26611616ab5616f/activerecord/lib/active_record/associations/collection_association.rb#L174-L179

which then calls the `destroy` on the collection association

https://github.com/rails/rails/blob/3dff30c14fa25f211088d62da26611616ab5616f/activerecord/lib/active_record/associations/collection_association.rb#L174-L179

and the finally it will call `remove_records` which will call the callbacks

https://github.com/rails/rails/blob/3dff30c14fa25f211088d62da26611616ab5616f/activerecord/lib/active_record/associations/collection_association.rb#L393-L403

### Why doesn't firm.clients.destroy work?

When we call `destroy` on this association we call the `destory` method in `ActiveRecord::Assocations::CollectionProxy` 

https://github.com/rails/rails/blob/02e8e6ebd4508f8474b11836f405da10cc59228f/activerecord/lib/active_record/associations/collection_proxy.rb#L690-L692

then calls the `destroy` method in `ActiveRecord::Associations::CollectionAssociation`, similar to the previous call when we did `firm.destroy`. However the main difference here is that `*records` in this case is empty

https://github.com/rails/rails/blob/3dff30c14fa25f211088d62da26611616ab5616f/activerecord/lib/active_record/associations/collection_association.rb#L197-L199

When `records` is empty, we return early when we call `delete_or_destroy`, therefore not calling any of the callbacks

https://github.com/rails/rails/blob/86480e71faea7ed152efcba633f0011a7c21ee33/activerecord/lib/active_record/associations/collection_association.rb#L379-L380

This tells me that me the way we should be calling destroy on associations is the way that @sebasoga suggested, where we pass the object we want to remove to `destroy`.

### Why does firm.clients.destroy_all work?

As we know, this call stack is very similar to `firm.destroy`, this is why callbacks are triggered. Also, this should be the way we should remove all associations and not `firm.client.destroy` as we just saw above. 🎉 

### Why does after_add work?

This callback works because we are calling it on the association. In the test when we call 
```ruby
client = firm.clients.create! name: "Client"
```

we are adding the object to the association which triggers the callback, the callback can also be triggered if we do something like 

```ruby
client = Client.create! name: "Client"
firm.clients << client
```

by simply creating a client object like `client = Client.create! name: "Client"` this will _not_ trigger any callbacks.

I hope this clarifies this issue for folks, at the very least it did for me 😅.

**Comment by noahd1** [comment]
I just ran into this situation in my application, when I noticed that my `after_remove` callbacks weren't being triggered as I expected. This thread helped me, so thanks to everyone who contributed (@nickborromeo @MatheusRich especially).

The TLDR for me is that if you want an `after_remove` callback to be triggered when removing one item of an association collection, you need to use this form:

```firm.clients.destroy(client)```

Other ways of destroying the record (e.g. `firm.clients.last.destroy`, `client.destroy`) do not trigger it.

I'm happy to live to fight another day, but also thought I'd raise a concern that this behavior, purely as an API consumer, even after reading the explanation, is still a bit surprising. I think this is due to an expectation that Active Record Callbacks set that do not apply to Association Callbacks. 

In particular, I think of Active Record callbacks as very "opt out" behavior -- as long as you're in object land instead of SQL land, they'll fire. `update_attribute`? Fires callbacks. `touch`? Yup. You can skip them, but you have to be deliberate about it. Caution is urged when calling methods that skip callbacks. From The Rails Guide: 

> These methods should be used with caution, however, because important business rules and application logic may be kept in callbacks.

With Association Callbacks, `after_remove` will only be called if you issue the correct removal call. There's easy ways to remove an object from a collection, and have your Association Callback not fire. Should "important business rules and application logic" live in Association Callbacks? Based on the current behavior, I'd argue probably not.

Now, I get it, Association Callbacks ARE NOT Active Record Callbacks. But it's a tough distinction. Since we're still dealing with the addition and removal of Active Record objects, it feels like there's daylight for confusion to creep in.
